contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1994 @@
+            to_address: recipient.into_string(),
+            amount: vec![Coin { denom, amount }],
+        }
+        .into()),
+        Denom::Cw20(addr) => {
+            let cw20_msg = to_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                recipient: recipient.into_string(),
+                amount,
+            })?;
+            Ok(WasmMsg::Execute {
+                contract_addr: addr.into_string(),
+                msg: cw20_msg,
+                funds: vec![],
+            }
+            .into())
+        }
+    }
+}
+
+pub fn update_rewards(deps: &mut DepsMut, env: &Env, addr: &Addr) -> StdResult<()> {
+    let config = CONFIG.load(deps.storage)?;
+    let reward_per_token = get_reward_per_token(deps.as_ref(), env, &config.staking_contract)?;
+    REWARD_PER_TOKEN.save(deps.storage, &reward_per_token)?;
+
+    let earned_rewards = get_rewards_earned(
+        deps.as_ref(),
+        env,
+        addr,
+        reward_per_token,
+        &config.staking_contract,
+    )?;
+    PENDING_REWARDS.update::<_, StdError>(deps.storage, addr.clone(), |r| {
+        Ok(r.unwrap_or_default() + earned_rewards)
+    })?;
+
+    USER_REWARD_PER_TOKEN.save(deps.storage, addr.clone(), &reward_per_token)?;
+    LAST_UPDATE_BLOCK.save(deps.storage, &env.block.height)?;
+    Ok(())
+}
+
+pub fn get_reward_per_token(deps: Deps, env: &Env, staking_contract: &Addr) -> StdResult<Uint256> {
+    let reward_config = REWARD_CONFIG.load(deps.storage)?;
+    let total_staked = get_total_staked(deps, staking_contract)?;
+    let current_block = min(env.block.height, reward_config.period_finish);
+    let last_update_block = LAST_UPDATE_BLOCK.load(deps.storage).unwrap_or_default();
+    let prev_reward_per_token = REWARD_PER_TOKEN.load(deps.storage).unwrap_or_default();
+    let additional_reward_per_token = if total_staked == Uint128::zero() {
+        Uint256::zero()
+    } else {
+        let numerator = reward_config
+            .reward_rate
+            .full_mul(Uint128::from(current_block - last_update_block))
+            .full_mul(scale_factor());
+        let denominator = Uint512::from(total_staked);
+        let result = numerator.checked_div(denominator)?;
+
+        // try_into() shouldn't cause problems here because final results fits into Uint256
+        result
+            .checked_div(Uint512::from(scale_factor()))?
+            .try_into()
+            .unwrap()
+    };
+
+    Ok(prev_reward_per_token + additional_reward_per_token)
+}
+
+pub fn get_rewards_earned(
+    deps: Deps,
+    _env: &Env,
+    addr: &Addr,
+    reward_per_token: Uint256,
+    staking_contract: &Addr,
+) -> StdResult<Uint128> {
+    let _config = CONFIG.load(deps.storage)?;
+    let staked_balance = get_staked_balance(deps, staking_contract, addr)?;
+    let user_reward_per_token = USER_REWARD_PER_TOKEN
+        .load(deps.storage, addr.clone())
+        .unwrap_or_default();
+    let reward_factor = reward_per_token.checked_sub(user_reward_per_token)?;
```

Comment:
1. I stake some tokens at a reward per token rate of 2.
2. The reward period for those ends.
3. The rewards are updated with a reward per token of 1.
4. I attempt to claim my rewards and this overflows.

Reply:
Note: this only applies about the `reward_per_token` value not being monotonically increasing in the other comment isn't off base.

Reply:
it is monotonically increasing